README
REDIS Server SNMP Agent.
Copyright (c) 2011 Brice Figureau <brice.figureau@daysofwonder.com>

PREAMBLE
========

redis-snmp is a Redis Net-SNMP agent written in Perl, and using the Net-Snmp 
Perl bindings.

It connects to a redis server and returns information to Net-SNMP when needed.
It parses the Redis INFO command results.

This work is based on Baron Schwartz Mysql Cacti Templates, available at:
http://code.google.com/p/mysql-cacti-templates/

This software is released under the GPL v2.

REQUIREMENTS
============

This package requires Net-SNMP version 5.4.3 or better. Previous versions have bugs preventing
returning 64 bits MIBs objects correctly from a perl agentX.

INSTALL
=======

Debian
------

Please build the package:
$ dpkg-buildpackage -rfakeroot -uc- b

and install it
Then, modify /etc/default/redis-snmp

Other
-----

Modify the Makefile, and:
$ sudo make install

Then create an init.d startup script. You can look to the debian one
in debian/init.d

Install the MIB file in your mibs directory.

CONFIGURE
=========

Configure net-snmp
------------------

Add:
master agentx
to /etc/snmp/snmpd.conf, and restart snmpd
Make sure the redis-snmp agent is able to open the agentX socket (permissions
and location of said socket can be managed with the snmpd.conf agentXPerms
directive).

Configure redis-snmp
--------------------

Under debian, the daemon can be configured with /etc/default/redis-snmp.
On all platform, the configuration is done with command line arguments:

-h|--host HOST          connect to redis HOST
-P|--port PORT          port to connect to (default 6379)
-p|--password DBPASS   use DBPASS as password to connect to redis
--oid OID               registering OID
-i|--refresh            refresh interval in seconds

OPENNMS
=======

You'll find in the opennms directory the various configuration snippets
that can be used to let opennms graph the value reported by this agent.

datacollection-config.xml
-------------------------

Add the content of opennms/redis.datacollection-config.xml to
your datacollection-config.xml file.

Then add this group to your hosts:
<systemDef name="Net-Snmp">
  <sysoidMask>.1.3.6.1.4.1.8072.3.</sysoidMask>
  <collect>
    <includeGroup>mib2-host-resources-system</includeGroup>
    <includeGroup>mib2-host-resources-memory</includeGroup>
    <includeGroup>net-snmp-disk</includeGroup>
    <includeGroup>ucd-loadavg</includeGroup>
    <includeGroup>ucd-memory</includeGroup>
    <includeGroup>ucd-sysstat</includeGroup>
    <includeGroup>redis</includeGroup>
  </collect>
</systemDef>

And now, add the graph definitions from the opennms/redis.snmp-graph.properties file 
to your snmp-graph.properties.


debian/changelog
redis-snmp (0.1-1) unstable; urgency=low

  * Initial release

 -- brice <brice+debian@daysofwonder.com>  Thu, 20 Oct 2011 11:17:19 +0100



debian/init.d
#! /bin/sh
#
# Copyright (c) 2008, 2009 Brice Figureau
#
# This init script is based on:
# the skeleton	example file to build /etc/init.d/ scripts.
#		This file should be used to construct scripts for /etc/init.d.
#
#		Written by Miquel van Smoorenburg <miquels@cistron.nl>.
#		Modified for Debian 
#		by Ian Murdock <imurdock@gnu.ai.mit.edu>.
#
# Version:	@(#)skeleton  1.9  26-Feb-2001  miquels@cistron.nl
#
### BEGIN INIT INFO
# Provides:          redis-snmp
# Required-Start:    $syslog
# Required-Stop:     $syslog
# Should-Start:      $local_fs $remote_fs $network $named $time
# Should-Stop:       $local_fs $remote_fs $network $named $time
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Start and stop the redis database server snmp agent
# Description:       Controls the launch/stop of redis SNMP agent.
### END INIT INFO

PATH=/sbin:/bin:/usr/sbin:/usr/bin
DAEMON=/usr/sbin/redis-snmp
DAEMON2="/usr/bin/perl"
NAME=redis-snmp
DESC="Redis-Server SNMP Agent"
PID_FILE=/var/run/redis-snmp.pid

test -x $DAEMON || exit 0

. /lib/lsb/init-functions

# Include redis-snmp defaults if available
if [ -f /etc/default/redis-snmp ] ; then
	. /etc/default/redis-snmp
fi

set -e

START="--start --quiet --pidfile ${PID_FILE} --name ${NAME} --startas ${DAEMON}"
STOP="--stop --quiet --pidfile ${PID_FILE} --name ${NAME}"

case "$1" in
  start)
		echo -n "Starting $DESC: "
    if start-stop-daemon ${START} -- --daemon-pid ${PID_FILE} ${DAEMON_OPTS} >/dev/null ; then
            echo "${NAME}."
    else
            if start-stop-daemon --test ${DAEMON_OPTS} >/dev/null 2>&1; then
                    echo "(failed)."
                    exit 1
            else
                    echo "(already running)."
                    exit 0
            fi
    fi
    ;;
  stop)
		log_daemon_msg "Stopping $DESC" "$NAME"
		killproc -p ${PID_FILE} $DAEMON2
		log_end_msg $?
    ;;
  restart)
    $0 stop
    exec $0 start
    ;;
  force-reload)
    $0 restart
    ;;
  status)
    pidofproc -p ${PID_FILE} $DAEMON2 >/dev/null
    status=$?
    if [ $status -eq 0 ]; then
            log_success_msg "redis-snmp server is running."
    else
            log_failure_msg "redis-snmp server is not running."
    fi
    exit $status
    ;;
  *)
    echo "Usage: $0 start|stop|restart|status|force-reload"
    exit 1
    ;;
esac

exit 0


opennms/tools/array.pl
my @types = (
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 1 - 4
    'Gauge32', 'Counter64', 'Gauge32', 'Gauge32',  # 5 - 8
    'Gauge32', 'Gauge32', 'Gauge32', 'Gauge32',  # 9 - 12
    'Gauge32', 'Gauge32', 'Counter32', 'Counter32',  # 13 - 16
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 17 - 20
    'Counter32', 'Gauge32', 'Gauge32', 'Gauge32',  # 21 - 24
    'Gauge32', 'Gauge32', 'Gauge32', 'Gauge32',  # 25 - 28
    'Gauge32', 'Gauge32', 'Counter32', 'Counter32',  # 29 - 32
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 33 - 36
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 37 - 40
    'Counter32', 'Counter32', 'Counter32', 'Gauge32',  # 41 - 44
    'Gauge32', 'Counter32', 'Counter32', 'Counter32',  # 45 - 48
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 49 - 52
    'Counter32', 'Gauge32', 'Gauge32', 'Counter32',  # 53 - 56
    'Gauge32', 'Gauge32', 'Gauge32', 'Counter32',  # 57 - 60
    'Gauge32', 'Gauge32', 'Counter32', 'Gauge32',  # 61 - 64
    'Gauge32', 'Gauge32', 'Gauge32', 'Counter32',  # 65 - 68
    'Counter32', 'Counter32', 'Counter32', 'Gauge32',  # 69 - 72
    'Gauge32', 'Counter32', 'Counter32', 'Counter32',  # 73 - 76
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 77 - 80
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 81 - 84
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 85 - 88
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 89 - 92
    'Counter32', 'Counter32', 'Counter32', 'Counter32',  # 93 - 96
    'Counter32', 'Counter64', 'Counter64', 'Gauge32',  # 97 - 100
    'Gauge32', 'Counter64', 'Counter64', 'Counter32',  # 101 - 104
    'Gauge32', 'Counter32', 'Counter32', 'Counter32',  # 105 - 108
    'Gauge32', 'Gauge32', 'Gauge32', 'Gauge32',  # 109 - 112
    'Gauge32', 'Gauge32', 'Gauge32', 'Gauge32',  # 113 - 116
    'Gauge32', 'Gauge32', 'Gauge32', 'Gauge32',  # 117 - 120
    'Gauge32', 'Gauge32', 'Gauge32', 'Gauge32',  # 121 - 124
    'Counter64', 'Counter64', 'Gauge32', 'Gauge32',  # 125 - 128
    'Gauge32', 'Gauge32', 'Gauge32', 'Gauge32',  # 129 - 132
    'Gauge32', 'Gauge32', 'Gauge32', 'Gauge32',  # 133 - 136
    'Counter64', 'Counter64', 'Counter64', 'Counter64',  # 137 - 140
    'Counter64', 'Counter64', 'Counter64',  # 141 - 143
);

my @newkeys = (
    'myKeyReadRequests', 'myKeyReads',  # 1 - 2
    'myKeyWriteRequests', 'myKeyWrites',  # 3 - 4
    'myHistoryList', 'myInnodbTransactions',  # 5 - 6
    'myReadViews', 'myCurrentTransactions',  # 7 - 8
    'myLockedTransactions', 'myActiveTransactions',  # 9 - 10
    'myPoolSize', 'myFreePages',  # 11 - 12
    'myDatabasePages', 'myModifiedPages',  # 13 - 14
    'myPagesRead', 'myPagesCreated',  # 15 - 16
    'myPagesWritten', 'myFileFsyncs',  # 17 - 18
    'myFileReads', 'myFileWrites',  # 19 - 20
    'myLogWrites', 'myPendingAIOLogIOs',  # 21 - 22
    'myPendingAIOSyncIOs', 'myPendingBufPoolFlushes',  # 23 - 24
    'myPendingChkpWrites', 'myPendingIbufAIOReads',  # 25 - 26
    'myPendingLogFlushes', 'myPendingLogWrites',  # 27 - 28
    'myPendingNormalAIOReads', 'myPendingNormalAIOWrites',  # 29 - 30
    'myIbufInserts', 'myIbufMerged',  # 31 - 32
    'myIbufMerges', 'mySpinWaits',  # 33 - 34
    'mySpinRounds', 'myOsWaits',  # 35 - 36
    'myRowsInserted', 'myRowsUpdated',  # 37 - 38
    'myRowsDeleted', 'myRowsRead',  # 39 - 40
    'myTableLocksWaited', 'myTableLocksImmediate',  # 41 - 42
    'mySlowQueries', 'myOpenFiles',  # 43 - 44
    'myOpenTables', 'myOpenedTables',  # 45 - 46
    'myInnodbOpenFiles', 'myOpenFilesLimit',  # 47 - 48
    'myTableCache', 'myAbortedClients',  # 49 - 50
    'myAbortedConnects', 'myMaxUsedConnections',  # 51 - 52
    'mySlowLaunchThreads', 'myThreadsCached',  # 53 - 54
    'myThreadsConnected', 'myThreadsCreated',  # 55 - 56
    'myThreadsRunning', 'myMaxConnections',  # 57 - 58
    'myThreadCacheSize', 'myConnections',  # 59 - 60
    'mySlaveRunning', 'mySlaveStopped',  # 61 - 62
    'mySlaveRetriedTransactions', 'mySlaveLag',  # 63 - 64
    'mySlaveOpenTempTables', 'myQcacheFreeBlocks',  # 65 - 66
    'myQcacheFreeMemory', 'myQcacheHits',  # 67 - 68
    'myQcacheInserts', 'myQcacheLowmemPrunes',  # 69 - 70
    'myQcacheNotCached', 'myQcacheQueriesInCache',  # 71 - 72
    'myQcacheTotalBlocks', 'myQueryCacheSize',  # 73 - 74
    'myQuestions', 'myComUpdate',  # 75 - 76
    'myComInsert', 'myComSelect',  # 77 - 78
    'myComDelete', 'myComReplace',  # 79 - 80
    'myComLoad', 'myComUpdateMulti',  # 81 - 82
    'myComInsertSelect', 'myComDeleteMulti',  # 83 - 84
    'myComReplaceSelect', 'mySelectFullJoin',  # 85 - 86
    'mySelectFullRangeJoin', 'mySelectRange',  # 87 - 88
    'mySelectRangeCheck', 'mySelectScan',  # 89 - 90
    'mySortMergePasses', 'mySortRange',  # 91 - 92
    'mySortRows', 'mySortScan',  # 93 - 94
    'myCreatedTmpTables', 'myCreatedTmpDiskTables',  # 95 - 96
    'myCreatedTmpFiles', 'myBytesSent',  # 97 - 98
    'myBytesReceived', 'myInnodbLogBufferSize',  # 99 - 100
    'myUnflushedLog', 'myLogBytesFlushed',  # 101 - 102
    'myLogBytesWritten', 'myRelayLogSpace',  # 103 - 104
    'myBinlogCacheSize', 'myBinlogCacheDiskUse',  # 105 - 106
    'myBinlogCacheUse', 'myBinaryLogSpace',  # 107 - 108
    'myStateClosingTables', 'myStateCopyingToTmpTable',  # 109 - 110
    'myStateEnd', 'myStateFreeingItems',  # 111 - 112
    'myStateInit', 'myStateLocked',  # 113 - 114
    'myStateLogin', 'myStatePreparing',  # 115 - 116
    'myStateReadingFromNet', 'myStateSendingData',  # 117 - 118
    'myStateSortingResult', 'myStateStatistics',  # 119 - 120
    'myStateUpdating', 'myStateWritingToNet',  # 121 - 122
    'myStateNone', 'myStateOther',  # 123 - 124
    'myAdditionalPoolAlloc', 'myTotalMemAlloc',  # 125 - 126
    'myHashIndexCellsTotal', 'myHashIndexCellsUsed',  # 127 - 128
    'myInnoDBLockStructs', 'myInnoDBLockWaitSecs',  # 129 - 130
    'myInnoDBTablesInUse', 'myInnoDBLockedTables',  # 131 - 132
    'myUncheckpointedBytes', 'myIBufCellCount',  # 133 - 134
    'myIBufUsedCells', 'myIBufFreeCells',  # 135 - 136
    'myAdaptiveHashMemory', 'myPageHashMemory',  # 137 - 138
    'myDictionaryCacheMemory', 'myFileSystemMemory',  # 139 - 140
    'myLockSystemMemory', 'myRecoverySystemMemory',  # 141 - 142
    'myThreadHashMemory',  # 143 - 143
);

my @oldkeys = (
    'Key_read_requests', 'Key_reads',  # 1 - 2
    'Key_write_requests', 'Key_writes',  # 3 - 4
    'history_list', 'innodb_transactions',  # 5 - 6
    'read_views', 'current_transactions',  # 7 - 8
    'locked_transactions', 'active_transactions',  # 9 - 10
    'pool_size', 'free_pages',  # 11 - 12
    'database_pages', 'modified_pages',  # 13 - 14
    'pages_read', 'pages_created',  # 15 - 16
    'pages_written', 'file_fsyncs',  # 17 - 18
    'file_reads', 'file_writes',  # 19 - 20
    'log_writes', 'pending_aio_log_ios',  # 21 - 22
    'pending_aio_sync_ios', 'pending_buf_pool_flushes',  # 23 - 24
    'pending_chkp_writes', 'pending_ibuf_aio_reads',  # 25 - 26
    'pending_log_flushes', 'pending_log_writes',  # 27 - 28
    'pending_normal_aio_reads', 'pending_normal_aio_writes',  # 29 - 30
    'ibuf_inserts', 'ibuf_merged',  # 31 - 32
    'ibuf_merges', 'spin_waits',  # 33 - 34
    'spin_rounds', 'os_waits',  # 35 - 36
    'rows_inserted', 'rows_updated',  # 37 - 38
    'rows_deleted', 'rows_read',  # 39 - 40
    'Table_locks_waited', 'Table_locks_immediate',  # 41 - 42
    'Slow_queries', 'Open_files',  # 43 - 44
    'Open_tables', 'Opened_tables',  # 45 - 46
    'innodb_open_files', 'open_files_limit',  # 47 - 48
    'table_cache', 'Aborted_clients',  # 49 - 50
    'Aborted_connects', 'Max_used_connections',  # 51 - 52
    'Slow_launch_threads', 'Threads_cached',  # 53 - 54
    'Threads_connected', 'Threads_created',  # 55 - 56
    'Threads_running', 'max_connections',  # 57 - 58
    'thread_cache_size', 'Connections',  # 59 - 60
    'slave_running', 'slave_stopped',  # 61 - 62
    'Slave_retried_transactions', 'slave_lag',  # 63 - 64
    'Slave_open_temp_tables', 'Qcache_free_blocks',  # 65 - 66
    'Qcache_free_memory', 'Qcache_hits',  # 67 - 68
    'Qcache_inserts', 'Qcache_lowmem_prunes',  # 69 - 70
    'Qcache_not_cached', 'Qcache_queries_in_cache',  # 71 - 72
    'Qcache_total_blocks', 'query_cache_size',  # 73 - 74
    'Questions', 'Com_update',  # 75 - 76
    'Com_insert', 'Com_select',  # 77 - 78
    'Com_delete', 'Com_replace',  # 79 - 80
    'Com_load', 'Com_update_multi',  # 81 - 82
    'Com_insert_select', 'Com_delete_multi',  # 83 - 84
    'Com_replace_select', 'Select_full_join',  # 85 - 86
    'Select_full_range_join', 'Select_range',  # 87 - 88
    'Select_range_check', 'Select_scan',  # 89 - 90
    'Sort_merge_passes', 'Sort_range',  # 91 - 92
    'Sort_rows', 'Sort_scan',  # 93 - 94
    'Created_tmp_tables', 'Created_tmp_disk_tables',  # 95 - 96
    'Created_tmp_files', 'Bytes_sent',  # 97 - 98
    'Bytes_received', 'innodb_log_buffer_size',  # 99 - 100
    'unflushed_log', 'log_bytes_flushed',  # 101 - 102
    'log_bytes_written', 'relay_log_space',  # 103 - 104
    'binlog_cache_size', 'Binlog_cache_disk_use',  # 105 - 106
    'Binlog_cache_use', 'binary_log_space',  # 107 - 108
    'State_closing_tables', 'State_copying_to_tmp_table',  # 109 - 110
    'State_end', 'State_freeing_items',  # 111 - 112
    'State_init', 'State_locked',  # 113 - 114
    'State_login', 'State_preparing',  # 115 - 116
    'State_reading_from_net', 'State_sending_data',  # 117 - 118
    'State_sorting_result', 'State_statistics',  # 119 - 120
    'State_updating', 'State_writing_to_net',  # 121 - 122
    'State_none', 'State_other',  # 123 - 124
    'additional_pool_alloc', 'total_mem_alloc',  # 125 - 126
    'hash_index_cells_total', 'hash_index_cells_used',  # 127 - 128
    'innodb_lock_structs', 'innodb_lock_wait_secs',  # 129 - 130
    'innodb_tables_in_use', 'innodb_locked_tables',  # 131 - 132
    'uncheckpointed_bytes', 'ibuf_cell_count',  # 133 - 134
    'ibuf_used_cells', 'ibuf_free_cells',  # 135 - 136
    'adaptive_hash_memory', 'page_hash_memory',  # 137 - 138
    'dictionary_cache_memory', 'file_system_memory',  # 139 - 140
    'lock_system_memory', 'recovery_system_memory',  # 141 - 142
    'thread_hash_memory',  # 143 - 143
);



opennms/tools/cacti2MIB.pl
# redis-snmp: cacti2MIB.pl
#
# This file contains the mapping from MySQL cacti templates to the MIB names, value and
# and types. This file is used to generate the opennms configuration.
#
# (c) Copryright 2011 - Brice Figureau
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.    See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

{   name        => 'Cacti 2 MIB',
    short_names => {
        'REDIS_connected_clients'           => { mib => 'redisConnectedClients',      order => 1, type => 'Gauge32' },
        'REDIS_connected_slaves'            => { mib => 'redisConnectedSlaves',       order => 2, type => 'Gauge32' },
        'REDIS_used_memory'                 => { mib => 'redisUsedMemory',            order => 3, type => 'Gauge64' },
        'REDIS_changes_since_last_save'     => { mib => 'redisChangesSinceLastSave',  order => 4, type => 'Gauge32' },
        'REDIS_total_connections_received'  => { mib => 'redisTotalConnections',      order => 5, type => 'Counter64' },
        'REDIS_total_commands_processed'    => { mib => 'redisCommandsProcessed',     order => 6, type => 'Counter64' },
    },
    startoid => '.1.3.6.1.4.1.20267.400.1',
}


opennms/tools/cacti2opennms.pl
#!/usr/bin/perl
#
# This is a script that generates OpenNMS config files from the MySQL Cacti
# Templates redis_definitions.pl file and the cacti to mib mapping
#
# Usage:
#   perl cacti2opennms.pl redis_definitions.pl cacti2MIB.pl
#
#
# This script is loosely based on make-templates.pl written by Baron Schwartz 
# for the MySQL Cacti Templates.
#
# Here is the original file comment
# This is a script that generates Cacti templates from a Perl data structure.
#
# This program is copyright (c) 2009 Brice Figureau. Feedback and improvements
# are welcome.
#
# THIS PROGRAM IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR IMPLIED
# WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED WARRANTIES OF
# MERCHANTIBILITY AND FITNESS FOR A PARTICULAR PURPOSE.
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of the GNU General Public License as published by the Free Software
# Foundation, version 2.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, write to the Free Software Foundation, Inc., 59 Temple
# Place, Suite 330, Boston, MA  02111-1307  USA.

# notes:
# data_source_type_id = 1 = "GAUGE"
# data_source_type_id = 2 = "COUNTER"
# data_source_type_id = 3 = "DERIVE"
# data_source_type_id = 4 = "ABSOLUTE"

use Data::Dumper;
use strict;
use warnings FATAL => 'all';
use Getopt::Long qw(:config auto_help auto_version no_ignore_case);
use Pod::Usage;

our $VERSION = '1.0';

use English qw(-no_match_vars);
use List::Util qw(max);

sub VersionMessage {
    print "cacti2opennms.pl $VERSION\n";
}

sub read_hash {
    my $filename = shift;
    open my $fh, '<', $filename or die "Can't open $filename: $OS_ERROR";
    my $code = do { local $INPUT_RECORD_SEPARATOR = undef; <$fh>; };
    close $fh;

    # This should be a hashref now.
    my $t = eval($code);
    if ( $EVAL_ERROR ) {
       die $EVAL_ERROR;
    }

    return $t;
}

my %opt = (
    graph_width => 565,
    graph_height => 200,
    output => "graph"
);

GetOptions(
    \%opt,
    "graph_height|h=i",
    "graph_width|w=i",
    "output|o=s",
    'man',
    'usage',
    "version|V" => sub {VersionMessage(); exit();},
) or pod2usage(-verbose => 0);

pod2usage(-verbose => 0) if $opt{usage};
pod2usage(-verbose => 1) if $opt{help};
pod2usage(-verbose => 2) if $opt{man};

if (scalar(@ARGV) != 2) {
    print "Missing cacti templates definitions file and/or cacti to mib mapping file\n";
    pod2usage(-verbose => 0);
}

# read redis template
my $t = read_hash(shift @ARGV);

# now read the cacti to MIB translation hash
my $cacti2MIB = read_hash(shift @ARGV);

my @mibKeysInOrder = map { $cacti2MIB->{short_names}->{$_}->{mib} } ( sort { $cacti2MIB->{short_names}->{$a}->{order} <=> $cacti2MIB->{short_names}->{$b}->{order} } keys %{ $cacti2MIB->{short_names} });
my %mibTypes = ();
my %mib2Cacti = ();

foreach my $cactiKey (keys %{ $cacti2MIB->{short_names} }) {
    my $mib = $cacti2MIB->{short_names}->{$cactiKey}->{mib};
    my $type = $cacti2MIB->{short_names}->{$cactiKey}->{type};

    $mibTypes{$mib} = $type;
    $mib2Cacti{$mib} = $cactiKey;
}

# Turn A_string_of_words into A String Of Words.
sub to_words {
   my ( $text ) = @_;
   return join(' ', map { ucfirst } split(/_/, $text));
}

# Removes vowels and compacts repeated letters to shorten text.  In this case,
# to 19 characters, which is RRDTool's internal limit.  This lets the data
# source (script) output long variable names, which can then be used as nice
# descriptive labels, while translating them into crunched text when needed for
# an RRA.
sub crunch {
   my $text = shift;
   my $cdef = shift;
   my $len = $cdef ? 18 : 19;

   return $text if $len && length($text) <= $len;
   $text = reverse $text; # work from the end backwards
   1 while ( length($text) > $len && $text =~ s/(?<![_ ])[aeiou]// );
   1 while ( length($text) > $len && $text =~ s/(.)\1+/$1/ );
   $text = reverse $text;
   die "Can't shorten $text enough" if length($text) > $len;
   return $text;
}

sub condense {
    my ( $text ) = @_;
    $text =~ s/ //g;
    $text =~ s/[^a-zA-Z0-9]//g;
    return lc($text);
}

sub emit_report {
    my $g = shift;
    my $rname = shift;
    my $list = shift;

   print <<END;
report.redis.${rname}.name=$g->{name}
report.redis.${rname}.columns=$list
report.redis.${rname}.type=nodeSnmp
report.redis.${rname}.width=$opt{graph_width}
report.redis.${rname}.height=$opt{graph_height}
report.redis.${rname}.command=--title "$g->{title}" \\
 --width $opt{graph_width} \\
 --height $opt{graph_height} \\
END
}

sub emit_defs {
    my $g = shift;
    my $names = shift;

    my $index = 0;
    foreach my $it ( @{ $g->{items} } ) {
        my $rrName = crunch($cacti2MIB->{short_names}->{$it->{item}}->{mib});
        my $varName = crunch($cacti2MIB->{short_names}->{$it->{item}}->{mib});
        $index=$index+1;
        print <<END;
 DEF:$rrName={rrd$index}:$varName:AVERAGE \\
END
    }
}

sub emit_rrd {
    my $name = shift;
    my $type = shift;
    my $color = shift;
    my $text = shift;
    print <<END;
 $type:$name#$color:"$text" \\
END
}

sub emit_cdef {
    my $name = shift;

    print <<END;
 CDEF:${name}c=${name},-1,* \\
END
    return "${name}c";
}

sub emit_gprint {
    my $name = shift;

    print <<END;
 GPRINT:${name}:AVERAGE:"Avg \\\\: %8.2lf %s" \\
 GPRINT:${name}:MIN:"Min \\\\: %8.2lf %s" \\
 GPRINT:${name}:MAX:"Max \\\\: %8.2lf %s\\\\n" \\
END
}

sub emit_last_gprint {
    my $name = shift;

    print <<END;
 GPRINT:${name}:AVERAGE:"Avg \\\\: %8.2lf %s" \\
 GPRINT:${name}:MIN:"Min \\\\: %8.2lf %s" \\
 GPRINT:${name}:MAX:"Max \\\\: %8.2lf %s\\\\n"
END
}

sub emit_datacollection {
    print <<END;
    <!-- REDIS-SERVER MIB -->
        <group name="redis" ifType="ignore">
END
    my $i = 0;
    foreach my $mib ( @mibKeysInOrder )
    {
        $i++;
        my $name = crunch($mib);
        my $oid = $cacti2MIB->{startoid}.".$i";
        my $type = $mibTypes{$mib};
        print <<END;
            <mibObj oid="$oid" instance="0" alias="$name" type="$type" />
END
    }
    print <<END;
        </group>
END
}

sub emit_report_list {
    my %reports = ();
LOOP:
    foreach my $g ( @{ $t->{graphs} } ) {
        # skip graphs we don't (yet) support
        my $it;
        foreach $it ( @{ $g->{items} } ) {
            my $item = $it->{item};
            unless(defined($cacti2MIB->{short_names}->{$item})) {
                print "skipping unknown graph: $it->{item}\n";
                next LOOP;
            }
            my $rname = condense($g->{name});
            $reports{$rname} = '';
        }
    }
    print join(', ', map { "redis.".$_ } keys %reports) . " \\\n";
    print "\n\n";
}

sub emit_graphs {
LOOP:
    foreach my $g ( @{ $t->{graphs} } ) {
        # skip graphs we don't (yet) support
        my $it;
        foreach $it ( @{ $g->{items} } ) {
            my $item = $it->{item};
            unless(defined($cacti2MIB->{short_names}->{$item})) {
                print "skipping unknown graph: $it->{item}\n";
                next LOOP;
            }
        }

        # Set the graph's title
        my $name = $g->{name};
        my $rname = condense($g->{name});
        my @res = ();
        $g->{title} = "$g->{name}";

        foreach $it ( @{ $g->{items} } ) {
            push(@res, crunch($cacti2MIB->{short_names}->{$it->{item}}->{mib}));
        }

        emit_report($g, $rname, join(',', @res));
        emit_defs($g);

        my $max = 0;
        foreach $it ( @{ $g->{items} } ) {
            my $text = to_words($it->{item});
            $max = length($text) if (length($text) > $max);
        }

        my $index = 0;
        foreach $it ( @{ $g->{items} } ) {
            unless(defined($cacti2MIB->{short_names}->{$it->{item}})) {
                next;
            }
            my $rrName = crunch($cacti2MIB->{short_names}->{$it->{item}}->{mib}, defined($it->{cdef}));
            my $type = $it->{type} || 'LINE1';
            my $t = to_words($it->{item});
            my $text = $t . (' ' x ($max - length($t)));

            my $rrNamec = defined($it->{cdef}) ? emit_cdef($rrName) : $rrName;
            emit_rrd($rrNamec, $type, $it->{color}, $text);
            if ($index++ < scalar @{ $g->{items}} -1) {
                emit_gprint($rrName);
            } else {
                emit_last_gprint($rrName);
            }
        }
        print "\n\n";
    }
}

sub array_print {
    my $columns = shift;
    my @array = @_;
    my $length = scalar @array;

    my $k = 0;
    while($length > 0) {
        my $str = '';
        my $rest = ($length <= $columns ? $length : $columns);
        for(my $i = 0; $i < $rest; $i++) {
            $str .= "'$array[$k + $i]', ";
        }
        $str .= " # ". ($k+1)." - ". ($k+$rest) ."\n";
        print ' ' x 4 . $str;
        $k += $rest;
        $length -= $rest;
    }
}

my %snmpTypes = (
    'Gauge32' => 'Gauge32',
    'Counter32' => 'Counter32',
    'Counter64' => 'Counter64',
    'Gauge64' => 'Counter64', # there is no gauge64 in SNMP
);

sub emit_agent {
    print "my \@types = (\n";
    array_print(4, map { $snmpTypes{$mibTypes{$_}} } @mibKeysInOrder);
    print ");\n\n";

    print "my \@newkeys = (\n";
    array_print(2, @mibKeysInOrder);
    print ");\n\n";

    print "my \@oldkeys = (\n";
    array_print(2, map { my $n = $mib2Cacti{$_} ; $n =~ s/^REDIS_//; $n } @mibKeysInOrder);
    print ");\n\n";
}

# Do the work.
# #############################################################################

# Graph templates
if ($opt{output} eq "graph") {
    emit_graphs();
}
elsif ($opt{output} eq "graphlist") {
    emit_report_list();
}
elsif ($opt{output} eq "datacollection") {
    emit_datacollection();
}
elsif ($opt{output} eq "agent") {
    emit_agent();
}
else {
    print "Unknown output format\n";
    pod2usage(-verbose => 0);
}


__END__

=head1 NAME

    cacti2opennms - create opennms configuration from Redis Cacti Templates 

=head1 SYNOPSIS

    cacti2opennms [options] redis_definitions.pl cacti2mib.pl

    -w, --graph_width WIDTH          width of graph in pixel
    -h, --graph_height HEIGHT        height of graph in pixel
    -o format                        which configuration to output
    -?, --help                       display this help and exit
    --usage                          display detailed usage information
    --man                            display program man page
    -V, --version                    output version information and exit

=head1 OPTIONS

=over 8

=item B<-w WIDTH,--graph_width WIDTH>

Graphs will be output with WIDTH pixels. Default value 565 pixels.

=item B<-h HEIGHT,--graph_height HEIGHT>

Graphs will be output with HEIGHT pixels. Default value 200 pixels.

=item B<-o format,--output format>

What kind of opennms configuration file to output.
Possible choices are:
  * graph: outputs the content of the snmp-graph.properties
  * graphlist: outputs the content of the snmp-graph.properties reports key
  * datacollection: outputs the content of the datacollections-config.xml file
  * agent: outputs the needed array for redis-snmp

=item B<--man>

Prints the manual page and exits.

=item B<--usage>

Prints detailed usage information and exits.

=item B<-?, --help>

Print a brief help message and exits.

=item B<-V, --version>

output version information and exit

=back

=head1 DESCRIPTION

B<redis-snmp> is a small daemon that connects to a local snmpd daemon
to report statistics on a local or remote Redis server.

=cut
